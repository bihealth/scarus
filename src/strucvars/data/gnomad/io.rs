//! I/O code for gnomAD constraints data.

use std::{io::BufReader, path::Path};

use serde::{Deserialize, Deserializer, Serialize, Serializer};

/// Deserialize `Option::None` as `"NA"`.
///
/// cf. https://stackoverflow.com/a/56384732/84349
fn deserialize_option_na<'de, D, T: Deserialize<'de>>(
    deserializer: D,
) -> Result<Option<T>, D::Error>
where
    D: Deserializer<'de>,
{
    // We define a local enum type inside of the function because it is untagged, serde will
    // deserialize as the first variant that it can.
    #[derive(Deserialize)]
    #[serde(untagged)]
    enum MaybeNA<U> {
        // If it can be parsed as Option<T>, it will be..
        Value(Option<U>),
        // ... otherwise try parsing as a string.
        NAString(String),
    }

    // Deserialize into local enum.
    let value: MaybeNA<T> = Deserialize::deserialize(deserializer)?;
    match value {
        // If parsed as T or None, return that.
        MaybeNA::Value(value) => Ok(value),

        // Otherwise, if value is string an "n/a", return None (and fail if it is any other
        // string)
        MaybeNA::NAString(string) => {
            if string == "NA" {
                Ok(None)
            } else {
                Err(serde::de::Error::custom("Unexpected string"))
            }
        }
    }
}

/// Serialize `Option::None` as `"NA"`.
fn serialize_option_na<S, T>(x: &Option<T>, s: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    T: Serialize,
{
    match x {
        Some(x) => s.serialize_some(x),
        None => s.serialize_str("NA"),
    }
}

/// A record from the gnomAD constraints database.
#[serde_with::skip_serializing_none]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Record {
    /// The Ensembl gene ID.
    pub ensembl_gene_id: String,
    /// The NCBI gene ID.
    pub entrez_id: String,
    /// The HGNC gene symbol.
    pub gene_symbol: String,
    /// The expected number of loss-of-function variants.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub exp_lof: Option<f64>,
    /// The expected number of missense variants.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub exp_mis: Option<f64>,
    /// The expected number of synonymous variants.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub exp_syn: Option<f64>,
    /// The missense-related Z-score.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub mis_z: Option<f64>,
    /// The observed number of loss-of-function variants.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub obs_lof: Option<u32>,
    /// The observed number of missense variants.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub obs_mis: Option<u32>,
    /// The observed number of synonymous variants.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub obs_syn: Option<u32>,
    /// The loss-of-function observed/expected ratio.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub oe_lof: Option<f64>,
    /// The lower bound of the loss-of-function observed/expected ratio.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub oe_lof_lower: Option<f64>,
    /// The upper bound of the loss-of-function observed/expected ratio.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub oe_lof_upper: Option<f64>,
    /// The missense observed/expected ratio.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub oe_mis: Option<f64>,
    /// The lower bound of the missense observed/expected ratio.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub oe_mis_lower: Option<f64>,
    /// The upper bound of the missense observed/expected ratio.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub oe_mis_upper: Option<f64>,
    /// The synonymous observed/expected ratio.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub oe_syn: Option<f64>,
    /// The lower bound of the synonymous observed/expected ratio.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub oe_syn_lower: Option<f64>,
    /// The upper bound of the synonymous observed/expected ratio.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub oe_syn_upper: Option<f64>,
    /// The probability of loss-of-function intolerance (pLI score).
    #[serde(
        alias = "pLI",
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub pli: Option<f64>,
    /// The synonymous-related Z-score.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub syn_z: Option<f64>,
    /// The probability of loss-of-function intolerance (pLI score) from ExAC.
    #[serde(
        alias = "exac_pLI",
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub exac_pli: Option<f64>,
    /// The observed number of loss-of-function variants from ExAC.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub exac_obs_lof: Option<f64>,
    /// The expected number of loss-of-function variants from ExAC.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub exac_exp_lof: Option<f64>,
    /// The loss-of-function observed/expected ratio from ExAC.
    #[serde(
        serialize_with = "serialize_option_na",
        deserialize_with = "deserialize_option_na"
    )]
    pub exac_oe_lof: Option<f64>,
}

/// Load gnomAD constraints file.
///
/// # Arguments
///
/// * `path` - Path to gnomAD constraints file.
///
/// # Returns
///
/// gnomAD constraints.
///
/// # Errors
///
/// If anything goes wrong, it returns a generic `anyhow::Error`.
pub fn load_file<P>(path: P) -> Result<Vec<Record>, anyhow::Error>
where
    P: AsRef<Path>,
{
    // Construct reader and skip initial 5 lines.
    tracing::debug!("opening file: {:?}", path.as_ref());
    let reader = std::fs::File::open(path)
        .map_err(|e| anyhow::anyhow!("problem opening file: {}", e))
        .map(BufReader::new)?;

    tracing::debug!("reading records...");
    let mut csv_reader = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .has_headers(true)
        .flexible(true)
        .from_reader(reader);
    let mut result = Vec::new();
    for record in csv_reader.deserialize() {
        let record: Record =
            record.map_err(|e| anyhow::anyhow!("problem parsing record: {}", e))?;
        result.push(record);
    }
    tracing::debug!("read a total of {} records", result.len());

    Ok(result)
}

#[cfg(test)]
mod test {
    #[test]
    fn test_load_file() -> Result<(), anyhow::Error> {
        let genes = super::load_file("tests/data/strucvars/gnomad_constraints.tsv")?;

        assert_eq!(genes.len(), 18481);
        insta::assert_yaml_snapshot!(&genes[0..5]);

        Ok(())
    }
}
