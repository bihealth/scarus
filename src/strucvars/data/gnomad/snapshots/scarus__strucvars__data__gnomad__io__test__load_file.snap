---
source: src/strucvars/data/gnomad/io.rs
expression: "&genes[0..5]"
---
- ensembl_gene_id: ENSG00000004059
  entrez_id: "381"
  gene_symbol: ARF5
  exp_lof: 9.0614
  exp_mis: 104.51
  exp_syn: 42.345
  mis_z: 2.2772
  obs_lof: 0
  obs_mis: 39
  obs_syn: 37
  oe_lof: 0
  oe_lof_lower: 0
  oe_lof_upper: 0.33
  oe_mis: 0.37316
  oe_mis_lower: 0.288
  oe_mis_upper: 0.487
  oe_syn: 0.87377
  oe_syn_lower: 0.671
  oe_syn_upper: 1.149
  pli: 0.94298
  syn_z: 0.6457
  exac_pli: 0.234
  exac_obs_lof: 2
  exac_exp_lof: 7.0385
  exac_oe_lof: 0.28415
- ensembl_gene_id: ENSG00000003056
  entrez_id: "4074"
  gene_symbol: M6PR
  exp_lof: 15.53
  exp_mis: 155.34
  exp_syn: 57.136
  mis_z: 0.97915
  obs_lof: 6
  obs_mis: 121
  obs_syn: 47
  oe_lof: 0.38635
  oe_lof_lower: 0.211
  oe_lof_upper: 0.763
  oe_mis: 0.77893
  oe_mis_lower: 0.671
  oe_mis_upper: 0.906
  oe_syn: 0.8226
  oe_syn_lower: 0.65
  oe_syn_upper: 1.049
  pli: 0.0092215
  syn_z: 1.0541
  exac_pli: 0.71303
  exac_obs_lof: 2
  exac_exp_lof: 12.531
  exac_oe_lof: 0.15961
- ensembl_gene_id: ENSG00000004478
  entrez_id: "2288"
  gene_symbol: FKBP4
  exp_lof: 23.481
  exp_mis: 250.85
  exp_syn: 95.792
  mis_z: 1.2082
  obs_lof: 6
  obs_mis: 197
  obs_syn: 109
  oe_lof: 0.25553
  oe_lof_lower: 0.139
  oe_lof_upper: 0.504
  oe_mis: 0.78533
  oe_mis_lower: 0.698
  oe_mis_upper: 0.884
  oe_syn: 1.1379
  oe_syn_lower: 0.973
  oe_syn_upper: 1.334
  pli: 0.15539
  syn_z: -1.0608
  exac_pli: 0.92747
  exac_obs_lof: 2
  exac_exp_lof: 16.906
  exac_oe_lof: 0.1183
- ensembl_gene_id: ENSG00000003137
  entrez_id: "56603"
  gene_symbol: CYP26B1
  exp_lof: 16.349
  exp_mis: 328.51
  exp_syn: 150.65
  mis_z: 1.0492
  obs_lof: 1
  obs_mis: 275
  obs_syn: 161
  oe_lof: 0.061167
  oe_lof_lower: 0.021
  oe_lof_upper: 0.29
  oe_mis: 0.83711
  oe_mis_lower: 0.758
  oe_mis_upper: 0.925
  oe_syn: 1.0687
  oe_syn_lower: 0.939
  oe_syn_upper: 1.218
  pli: 0.97987
  syn_z: -0.66253
  exac_pli: 0.95435
  exac_obs_lof: 0
  exac_exp_lof: 9.4081
  exac_oe_lof: 0
- ensembl_gene_id: ENSG00000003509
  entrez_id: "55471"
  gene_symbol: NDUFAF7
  exp_lof: 23.494
  exp_mis: 237.41
  exp_syn: 85.64
  mis_z: -1.0514
  obs_lof: 27
  obs_mis: 283
  obs_syn: 100
  oe_lof: 1.1492
  oe_lof_lower: 0.846
  oe_lof_upper: 1.582
  oe_mis: 1.192
  oe_mis_lower: 1.081
  oe_mis_upper: 1.315
  oe_syn: 1.1677
  oe_syn_lower: 0.992
  oe_syn_upper: 1.379
  pli: 0.00000000000000000010818
  syn_z: -1.2198
  exac_pli: 0.000000000011773
  exac_obs_lof: 17
  exac_exp_lof: 17.154
  exac_oe_lof: 0.99105

